WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,273 @@
+        }
+        return cachesURL
+    }
+
+    /// Creates a cache instance with a given root URL.
+    public init(url: URL, configuration: Configuration = .init()) {
+        self.rootURL = url
+        self.configuration = configuration
+
+        do {
+            try FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+        } catch {
+            DDLogError("Failed to creates cache root directory at: \(url) with error: \(error)")
+        }
+        DispatchQueue.main.asyncAfter(deadline: .now() + configuration.initialSweepDelay) { [weak self] in
+            Task {
+                await self?.performAndScheduleNextSweep()
+            }
+        }
+    }
+
+    // MARK: Codable
+
+    public func getValue<T: Decodable>(
+        _ type: T.Type,
+        forKey key: String,
+        decoder: JSONDecoder = JSONDecoder()
+    ) async -> T? {
+        guard let data = getData(forKey: key) else { return nil }
+        return try? decoder.decode(type, from: data)
+    }
+
+    public func setValue<T: Encodable>(
+        _ value: T,
+        forKey key: String,
+        encoder: JSONEncoder = JSONEncoder()
+    ) async {
+        guard let data = try? encoder.encode(value) else { return }
+        setData(data, forKey: key)
+    }
+
+    public func removeValue(forKey key: String) {
+        removeData(forKey: key)
+    }
+
+    // MARK: Codable (Closures)
+
+    public nonisolated func getValue<T: Decodable>(
+        _ type: T.Type,
+        forKey key: String,
+        decoder: JSONDecoder = JSONDecoder(),
+        _ completion: @escaping (T?) -> Void
+    ) {
+        Task {
+            let value = await getValue(type, forKey: key, decoder: decoder)
+            completion(value)
+        }
+    }
+
+    public nonisolated func setValue<T: Encodable>(
+        _ value: T,
+        forKey key: String,
+        encoder: JSONEncoder = JSONEncoder()
+    ) {
+        Task {
+            await setValue(value, forKey: key, encoder: encoder)
+        }
+    }
+
+    // MARK: Accessing Cached Data
+
+    public func getData(forKey key: String) -> Data? {
+        guard let url = fileURL(for: key) else { return nil }
+        return try? Data(contentsOf: url)
+    }
+
+    public func setData(_ data: Data, forKey key: String) {
+        guard let url = fileURL(for: key) else { return }
+        do {
+            try data.write(to: url)
+        } catch let error as NSError {
+            guard error.code == CocoaError.fileNoSuchFile.rawValue && error.domain == CocoaError.errorDomain else { return }
+            try? FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+            try? data.write(to: url) // re-create a directory and try again
+        }
+    }
+
+    public func removeData(forKey key: String) {
+        guard let url = fileURL(for: key) else { return }
+        try? FileManager.default.removeItem(at: url)
```

Comment:
We shouldn't silently swallow this error – the caller should have to decide how to deal with it

Reply:
Hey, thanks for the review! 

I see a common theme in many of the comments: error handling. Not exposing errors to the user and using optionals was a deliberate choice. We don't expect any of these methods to fail, and if they do, I don't think there is any point in exposing the error to the caller. It's a cache, so we will never show errors to the user. The best you can do is log the error and move on, so this is what this class already does. Exposing the errors will make the API less ergonomic.

The initializer is also deliberately non-failing so that you could have a non-nill `DiskCache.shared` instance (it doesn't have to be a global variable).

It's designed similarly to other caches, e.g. native [`URLCache`](https://developer.apple.com/documentation/foundation/urlcache). Granted, this API is a bit old, but it never exposes errors to the user.

Reply:
> There is no error recovery that the caller can possibly perform.

Not necessarily – if writes to an on-disk cache are failing, we could examine whether I have other cached resources kicking around – cached reader resources, the `URLSession` cache. Alternatively, the issue could be that the user's disk is full (in which case we should probably avoid making that problem worse). But we won't know any of that if the error is never surfaced.

Additionally, if this cache were used somewhere like a widget, an inability to use a local cache would be great to have as one of the signals controlling update frequency (as every update would need to wake up the device's radios since we'd never have local values present to work from).

Lastly, if this cache is unusable but it's critically important that the data be retained, the developer could choose to use an in-memory cache instead (like with the `URLCache` example you provided). 

Just a couple off-the-top-of-my-head examples, but I'm not sure it's fair to say that there's nothing the user can do.

(also – if the error is that the user's disk is full, it'd be great to warn them about that on a cache miss, rather than after they've written a post and we can't persist it)

Reply:
I expanded my thoughts on this here a little bit https://github.com/wordpress-mobile/WordPress-iOS/pull/21059#issuecomment-1644287297.

> Lastly, if this cache is unusable but it's critically important that the data be retained

If that's the case, the cache should not be used. it makes no guarantees that it's available by design.

> ather than after they've written a post and we can't persist it)

That would be a programmer error. The cache should never be used for user data.